src/lib/FreeC/Pass/TypeInferencePass.hs

```diff
@@ -666,6 +666,22 @@ annotateExprWith' (IR.Lambda srcSpan args expr _) resType =
     addTypeEquation srcSpan funcType resType
     return (IR.Lambda srcSpan args' expr' (makeExprType resType))
 
+
+annotateExprWith' (IR.Let srcSpan binds expr _) resType =
+  withLocalTypeAssumption $ do
+    retType  <- liftConverter freshTypeVar
+    binds'   <- mapM annotateBind binds
+    expr'    <- annotateExprWith expr retType
+    return (IR.Let srcSpan binds' expr' (makeExprType resType))
+     where
+       annotateBind :: IR.Bind -> TypeInference IR.Bind
+       annotateBind (IR.Bind bindSrcSpan bindVarPat bindExpr) = do
+           bindVarPat' <- annotateVarPat bindVarPat
+           bindType    <- liftConverter freshTypeVar
+           bindExpr'   <- annotateExprWith bindExpr bindType
+           addTypeEquation srcSpan (fromJust $ IR.varPatType bindVarPat') bindType
```

Comment:
I guess that we can recycle the type the variable pattern has been annotated with.

```suggestion
           let Just bindType = IR.varPatType bindVarPat'
           bindExpr' <- annotateExprWith bindExpr bindType
```